crates/xline-client/src/clients/lock.rs

```diff
@@ -1,78 +1,247 @@
-            .field("lease_client", &self.lease_client)
-            .field("watch_client", &self.watch_client)
-            .field("token", &self.token)
-            .finish()
+    #[must_use]
+    pub fn new(client: LockClient) -> Self {
+        Self {
+            client,
+            lease_id: 0,
+            ttl: DEFAULT_SESSION_TTL,
+            keep_alive: None,
+        }
+    }
+
+    /// Set session TTL.
+    #[inline]
+    #[must_use]
+    pub const fn with_ttl(mut self, ttl: i64) -> Self {
+        self.ttl = ttl;
+        self
+    }
+
+    /// Set session lease.
+    #[inline]
+    #[must_use]
+    pub const fn with_lease(mut self, lease_id: i64) -> Self {
+        self.lease_id = lease_id;
+        self
+    }
+
+    /// Crate an auto-renew session.
+    ///
+    /// # Errors
+    /// return errors when lease client grant failed
+    #[inline]
+    pub async fn build(mut self) -> Result<Self> {
+        let ttl = self.ttl;
+        self.lease_id = if self.lease_id == 0 {
+            let lease_response = self
+                .client
+                .lease_client
+                .grant(LeaseGrantRequest::new(ttl))
+                .await?;
+            lease_response.id
+        } else {
+            self.lease_id
+        };
+        let mut lease_client = self.client.lease_client.clone();
+        let lease_id = self.lease_id;
+        self.keep_alive = Some(tokio::spawn(async move {
+            /// The renew interval factor of which value equals 60% of one second.
+            const RENEW_INTERVAL_FACTOR: u64 = 600;
+            let (mut keeper, mut stream) = lease_client
+                .keep_alive(LeaseKeepAliveRequest::new(lease_id))
+                .await?;
+            loop {
+                keeper.keep_alive()?;
+                if let Some(resp) = stream.message().await? {
+                    if resp.ttl < 0 {
+                        return Err(XlineClientError::InvalidArgs(String::from(
+                            "lease keepalive response has negative ttl",
+                        )));
+                    }
+                    sleep(Duration::from_millis(
+                        resp.ttl.unsigned_abs().overflow_mul(RENEW_INTERVAL_FACTOR),
+                    ))
+                    .await;
+                }
+            }
+        }));
+        Ok(self)
     }
 }
 
-// These methods primarily originate from xline lock server,
-// see also: `xline/src/server/lock_server.rs`
-impl LockClient {
-    /// Creates a new `LockClient`
+impl Drop for Session {
     #[inline]
-    pub fn new(
-        curp_client: Arc<CurpClient>,
-        channel: Channel,
-        token: Option<String>,
-        id_gen: Arc<LeaseIdGenerator>,
-    ) -> Self {
+    fn drop(&mut self) {
+        if let Some(keep_alive) = self.keep_alive.take() {
+            keep_alive.abort();
+        }
+    }
+}
+
+/// Xutex（Xline Mutex） implements the sync lock with xline
+#[derive(Debug)]
+pub struct Xutex {
```

Comment:
Should we follow the Rust mutex type style and make this `Xutex<T>`?

Reply:
The `Xutex` will only lock a string key. Then, what is the meaning of providing a generic parameter?

Reply:
As we can't guarantee the safety of the lock, I think we don't needs to implement this pattern anymore.